rudder-core/src/main/scala/com/normation/rudder/repository/jdbc/WorkflowJdbcRepository.scala

```diff
@@ -38,166 +38,73 @@ package com.normation.rudder.repository.jdbc
-    Try {
-      roRepo.isChangeRequestInWorkflow(crId) match {
-        case eb : EmptyBox => eb
-        case Full(true) =>
-          jdbcTemplate.update(
-              INSERT_SQL
-            , new java.lang.Integer(crId.value)
-            , state.value
-          )
-          roRepo.getStateOfChangeRequest(crId)
-        case _ =>
-          logger.error(s"Cannot start a workflow for Change Request id ${crId.value}, as it is already part of a workflow")
-          Failure(s"Cannot start a workflow for Change Request id ${crId.value}, as it is already part of a workflow")
-
+    val process = {
+      for {
+        exists  <- sql"""select state from workflow where id = ${crId}""".query[WorkflowNodeId].option
+        created <- exists match {
+                     case None    => sql"""insert into workflow (id, state) values (${crId},${state})""".update.run.attempt
+                     case Some(s) => (-\/(s"Cannot start a workflow for Change Request id ${crId.value}, "+
+                                     s"as it is already part of a workflow in state '${s}'")).pure[ConnectionIO]
+                   }
+      } yield {
+        state
       }
-    } match {
-      case Success(x) => x
-      case Catch(error) =>
-        logger.error(s"Error when updating status of change request ${crId.value}: ${error.toString}")
-        Failure(error.toString())
     }
+    process.attempt.transact(xa).unsafePerformSync
   }
 
   def updateState(crId: ChangeRequestId, from :  WorkflowNodeId, state : WorkflowNodeId) : Box[WorkflowNodeId] = {
-    Try {
-      roRepo.getStateOfChangeRequest(crId) match {
-        case eb : EmptyBox => eb
-        case Full(entry) =>
-          if (entry != from) {
-            Failure(s"Cannot change status of ChangeRequest id ${crId.value} : it has the status ${entry.value} but we were expecting ${from.value}")
-          } else {
-            jdbcTemplate.update(
-                  UPDATE_SQL
-                , state.value
-                , new java.lang.Integer(crId.value)
-              )
-          }
-          roRepo.getStateOfChangeRequest(crId)
+    val process = {
+      for {
+        exists  <- sql"""select state from workflow where id = ${crId}""".query[WorkflowNodeId].option
+        created <- exists match {
+                     case Some(s) =>
+                       if(s == from) {
+                         sql"""update workflow set state = ${state} where id = ${crId}""".update.run.attempt
+                       } else {
+                         (-\/(s"Cannot change status of ChangeRequest '${crId.value}': it has the status '${s.value}' "+
+                              s"but we were expecting '${from.value}'. Perhaps someone else changed it concurently?").pure[ConnectionIO])
+                       }
+                     case None    => (-\/(s"Cannot start a workflow for Change Request id ${crId.value}, "+
```

Comment:
wrong text: Cannot change status of workflow for Change Request id, as it is not part of any workflow yet